Support media-progress() and container-progress() CSS functions

weinig opened the pull request:
#### 79434c0efe51e2fc68ac9980f6f9a948dba0e58d
<pre>
Support media-progress() and container-progress() CSS functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=284669">https://bugs.webkit.org/show_bug.cgi?id=284669</a>

Reviewed by Darin Adler.

Adds support for the `media-progress()` and `container-progress()` CSS
functions from CSS Values and Units 5:

    <a href="https://drafts.csswg.org/css-values-5/#media-progress-func">https://drafts.csswg.org/css-values-5/#media-progress-func</a>
    <a href="https://drafts.csswg.org/css-values-5/#container-progress-func">https://drafts.csswg.org/css-values-5/#container-progress-func</a>

Also updates the progress() CSS function from the same spec to its
updated syntax, removing keyword separators in favor of commas.

`media-progress()` and `container-progress()` are  implemented behind
their own feature flags `CSSMediaProgressFunctionEnabled` and
`CSSContainerProgressFunctionEnabled` respectively, both of which
are set to the status level &quot;preview&quot;.

The bulk of the change is factoring the existing media query and
container query feature schemas to support alternative &quot;value
providing&quot; interfaces which rather than reporting a boolean match
result, return a double value in canonical units. The remainder
is adding the parsing/serialization/evaluation for the new calc
nodes for each function.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-interpolation-math-functions-tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-interpolation-math-functions-tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-math-functions-tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-math-functions-tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-serialize.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-serialize.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-serialize.tentative.html:
    - Updates tests to the latest syntax in the spec and update now passing results.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new features for the new functions.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords for `media-progress` and `container-progress`.

* Source/WebCore/css/calc/CSSCalcTree+ContainerProgressEvaluator.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+ContainerProgressEvaluator.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+MediaProgressEvaluator.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+MediaProgressEvaluator.h: Added.
    - Rather than duplicate evaluation in both the simplification and evaluation
      phases, put the core evaluation of the new nodes into their own files.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
    - Add support for the MediaProgress and ContainerProgress calc nodes.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add new settings for media-progress and container-progress for access
      in the calc parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.h:
    - Expose a helper to reject invalid container names; used by the
      calc parser for container-progress().

* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/css/query/ContainerQueryFeatures.h:
* Source/WebCore/css/query/ContainerQueryParser.cpp:
* Source/WebCore/css/query/ContainerQueryParser.h:
* Source/WebCore/css/query/GenericMediaQueryEvaluator.cpp:
* Source/WebCore/css/query/GenericMediaQueryEvaluator.h:
* Source/WebCore/css/query/GenericMediaQueryTypes.h:
* Source/WebCore/css/query/MediaQuery.h:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/css/query/MediaQueryFeatures.cpp:
* Source/WebCore/css/query/MediaQueryFeatures.h:
* Source/WebCore/css/query/MediaQueryParser.cpp:
* Source/WebCore/css/query/MediaQueryParser.h:
    - Refactor media query/container query schema code to allow
      schemas to conform to new MediaProgressProviding and its
      analog ContainerProgressProviding. Like the schemas, these
      are provided as type erased singleton instances accessed
      by name lookup.

Canonical link: <a href="https://commits.webkit.org/287862@main">https://commits.webkit.org/287862@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5868dd15ae29e0dcdb3209ae6ad0200fccbb0699

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81140 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/665 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35083 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85669 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32126 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83250 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/683 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8473 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63348 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21114 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/434 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73846 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43646 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/331 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28008 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30584 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/74119 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71855 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87104 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/80198 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5931 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71652 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8549 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69681 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70887 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17645 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14930 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13852 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/102605 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8331 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13854 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/24929 "Found 1 new JSC stress test failure: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-bbq (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8168 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11688 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9976 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
```diff
@@ -703,6 +710,192 @@ static std::optional<TypedChild> consumeProgress(CSSParserTokenRange& tokens, in
+        container = tokens.consumeIncludingWhitespace().value().toAtomString();
+    }
+
+    const auto* schema = CQ::ContainerQueryParser::containerProgressProvidingSchemaForFeatureName(*featureName, state.parserContext);
```
Is `const auto*` really better than `auto*` or `auto` here? Presumably the const-ness is already present in the return value and isn’t added by stating the type?

darinadler commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -728,20 +735,59 @@ struct Progress {
 };
 
+struct MediaProgress {
+    WTF_MAKE_TZONE_ALLOCATED(MediaProgress);
```
Don’t we have a `STRUCT` version of this macro that leaves us in public so we don’t have to write public out afterward?

weinig replied:
WTF_MAKE_STRUCT_TZONE_ALLOCATED!

darinadler commented on Source/WebCore/css/query/ContainerQueryParser.cpp:
```diff
@@ -76,16 +81,19 @@ const MQ::FeatureSchema* ContainerQueryParser::schemaForFeatureName(const AtomSt
+    using Map = MemoryCompactLookupOnlyRobinHoodHashMap<AtomString, const ContainerProgressProviding*>;
+
+    static NeverDestroyed<Map> schemas = [&] {
+        auto entries = Features::allContainerProgressProvidingSchemas();
```
Can we do this without a local variable?

darinadler commented on Source/WebCore/css/query/MediaQueryParser.cpp:
```diff
@@ -187,6 +187,21 @@ const FeatureSchema* MediaQueryParser::schemaForFeatureName(const AtomString& na
+    using Map = MemoryCompactLookupOnlyRobinHoodHashMap<AtomString, const MediaProgressProviding*>;
+
+    static NeverDestroyed<Map> schemas = [&] {
+        auto entries = Features::allMediaProgressProvidingSchemas();
```
Can we do this without a local variable?

weinig committed:
Support media-progress() and container-progress() CSS functions
https://bugs.webkit.org/show_bug.cgi?id=284669

Reviewed by Darin Adler.

Adds support for the `media-progress()` and `container-progress()` CSS
functions from CSS Values and Units 5:

    https://drafts.csswg.org/css-values-5/#media-progress-func
    https://drafts.csswg.org/css-values-5/#container-progress-func

Also updates the progress() CSS function from the same spec to its
updated syntax, removing keyword separators in favor of commas.

`media-progress()` and `container-progress()` are  implemented behind
their own feature flags `CSSMediaProgressFunctionEnabled` and
`CSSContainerProgressFunctionEnabled` respectively, both of which
are set to the status level "preview".

The bulk of the change is factoring the existing media query and
container query feature schemas to support alternative "value
providing" interfaces which rather than reporting a boolean match
result, return a double value in canonical units. The remainder
is adding the parsing/serialization/evaluation for the new calc
nodes for each function.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-interpolation-math-functions-tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/rotate-interpolation-math-functions-tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-math-functions-tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/scale-animation-math-functions-tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/container-progress-serialize.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/media-progress-serialize.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-computed.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-computed.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-invalid.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-invalid.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-serialize.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/progress-serialize.tentative.html:
    - Updates tests to the latest syntax in the spec and update now passing results.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
    - Add new features for the new functions.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files.

* Source/WebCore/css/CSSValueKeywords.in:
    - Add new keywords for `media-progress` and `container-progress`.

* Source/WebCore/css/calc/CSSCalcTree+ContainerProgressEvaluator.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+ContainerProgressEvaluator.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+MediaProgressEvaluator.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+MediaProgressEvaluator.h: Added.
    - Rather than duplicate evaluation in both the simplification and evaluation
      phases, put the core evaluation of the new nodes into their own files.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcTree.h:
    - Add support for the MediaProgress and ContainerProgress calc nodes.

* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
    - Add new settings for media-progress and container-progress for access
      in the calc parser.

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Conditional.h:
    - Expose a helper to reject invalid container names; used by the
      calc parser for container-progress().

* Source/WebCore/css/query/ContainerQueryFeatures.cpp:
* Source/WebCore/css/query/ContainerQueryFeatures.h:
* Source/WebCore/css/query/ContainerQueryParser.cpp:
* Source/WebCore/css/query/ContainerQueryParser.h:
* Source/WebCore/css/query/GenericMediaQueryEvaluator.cpp:
* Source/WebCore/css/query/GenericMediaQueryEvaluator.h:
* Source/WebCore/css/query/GenericMediaQueryTypes.h:
* Source/WebCore/css/query/MediaQuery.h:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/css/query/MediaQueryFeatures.cpp:
* Source/WebCore/css/query/MediaQueryFeatures.h:
* Source/WebCore/css/query/MediaQueryParser.cpp:
* Source/WebCore/css/query/MediaQueryParser.h:
    - Refactor media query/container query schema code to allow
      schemas to conform to new MediaProgressProviding and its
      analog ContainerProgressProviding. Like the schemas, these
      are provided as type erased singleton instances accessed
      by name lookup.

Canonical link: https://commits.webkit.org/287862@main
